src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4723,3 +4723,174 @@ void C2_MacroAssembler::vector_reverse_byte(BasicType bt, XMMRegister dst, XMMRe
+      vpternlogd(xtmp1, 0xff, xtmp1, xtmp1, vec_enc);
+      vpunpcklwd(xtmp2, xtmp1, src, vec_enc);
+      evplzcntd(xtmp2, k0, xtmp2, merge, vec_enc);
+      vpunpckhwd(dst, xtmp1, src, vec_enc);
+      evplzcntd(dst, k0, dst, merge, vec_enc);
+      vpackusdw(dst, xtmp2, dst, vec_enc);
+      break;
+    case T_BYTE:
+      // T1 = Compute leading zero counts of 4 LSB bits of each byte by
+      // accessing the lookup table.
+      // T2 = Compute leading zero counts of 4 MSB bits of each byte by
+      // accessing the lookup table.
+      // Add T1 to T2 if 4 MSB bits of byte are all zeros.
+      assert(VM_Version::supports_avx512bw(), "");
+      evmovdquq(xtmp1, ExternalAddress(StubRoutines::x86::vector_count_leading_zeros_lut()), vec_enc, rtmp);
+      movl(rtmp, 0x0F0F0F0F);
+      evpbroadcastd(xtmp2, rtmp, vec_enc);
+      vpand(dst, xtmp2, src, vec_enc);
+      vpshufb(dst, xtmp1, dst, vec_enc);
+      vpsrlw(xtmp3, src, 4, vec_enc);
+      vpand(xtmp3, xtmp2, xtmp3, vec_enc);
+      vpshufb(xtmp2, xtmp1, xtmp3, vec_enc);
+      vpxor(xtmp1, xtmp1, xtmp1, vec_enc);
+      evpcmpeqb(ktmp, xtmp1, xtmp3, vec_enc);
+      vpaddb(dst, dst, xtmp2, vec_enc);
+      evpblendmb(dst, ktmp, xtmp2, dst, true, vec_enc);
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+}
+void C2_MacroAssembler::vector_count_leading_zeros_byte_avx(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                                            XMMRegister xtmp2, XMMRegister xtmp3, Register rtmp, int vec_enc) {
+  vmovdqu(xtmp1, ExternalAddress(StubRoutines::x86::vector_count_leading_zeros_lut()));
+  movl(rtmp, 0x0F0F0F0F);
+  movdl(xtmp2, rtmp);
+  vpbroadcastd(xtmp2, xtmp2, vec_enc);
+  // T1 = Compute leading zero counts of 4 LSB bits of each byte by
+  // accessing the lookup table.
+  vpand(dst, xtmp2, src, vec_enc);
+  vpshufb(dst, xtmp1, dst, vec_enc);
+  // T2 = Compute leading zero counts of 4 MSB bits of each byte by
+  // accessing the lookup table.
+  vpsrlw(xtmp3, src, 4, vec_enc);
+  vpand(xtmp3, xtmp2, xtmp3, vec_enc);
+  vpshufb(xtmp2, xtmp1, xtmp3, vec_enc);
+  // Add T1 to T2 if 4 MSB bits of byte are all zeros.
+  vpxor(xtmp1, xtmp1, xtmp1, vec_enc);
+  vpcmpeqb(xtmp3, xtmp1, xtmp3, vec_enc);
+  vpaddb(dst, dst, xtmp2, vec_enc);
+  vpblendvb(dst, xtmp2, dst, xtmp3, vec_enc);
+}
+
+void C2_MacroAssembler::vector_count_leading_zeros_short_avx(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                                             XMMRegister xtmp2, XMMRegister xtmp3, Register rtmp, int vec_enc) {
+   vector_count_leading_zeros_byte_avx(dst, src, xtmp1, xtmp2, xtmp3, rtmp, vec_enc);
+   // Add zero counts of lower byte and upper byte of a word if
+   // upper byte holds a zero value.
+   vpsrlw(xtmp3, src, 8, vec_enc);
+   vpcmpeqw(xtmp3, xtmp1, xtmp3, vec_enc);
+   vpsllw(xtmp2, dst, 8, vec_enc);
+   vpaddw(xtmp2, xtmp2, dst, vec_enc);
+   vpblendvb(dst, dst, xtmp2, xtmp3, vec_enc);
```

Comment:
The mask is generated using a word operation, but blend is a byte operation?

Reply:
There is no variable word/double word/quadword blend,  generated mask lane is composed of multiple bytes and hence byte level blend is sufficient to blend multi-byte lanes. We do have a single/double precision blend but its latency is similar to byte level blend in addition it may incur domain switch over penalty.